pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -992,9 +1043,138 @@ func (r *ReconcileClusterPool) createCloudBuilder(pool *hivev1.ClusterPool, logg
 		cloudBuilder.Region = platform.Azure.Region
 		cloudBuilder.CloudName = platform.Azure.CloudName
 		return cloudBuilder, nil
-	// TODO: OpenStack, VMware, and Ovirt.
+	case platform.OpenStack != nil:
+		credsSecret, err := r.getCredentialsSecret(pool, platform.OpenStack.CredentialsSecretRef.Name, logger)
+		if err != nil {
+			return nil, err
+		}
+		cloudBuilder := clusterresource.NewOpenStackCloudBuilderFromSecret(credsSecret)
+		cloudBuilder.Cloud = platform.OpenStack.Cloud
+		return cloudBuilder, nil
+	// TODO: VMware, and Ovirt.
 	default:
 		logger.Info("unsupported platform")
 		return nil, errors.New("unsupported platform")
 	}
 }
+
+func (r *ReconcileClusterPool) getInventoryCustomization(pool *hivev1.ClusterPool, logger log.FieldLogger) (*corev1.LocalObjectReference, error) {
+	cdcRefs := pool.Spec.Inventory.ClusterDeploymentCustomizations
+	for _, ref := range cdcRefs {
+		cdc := &hivev1.ClusterDeploymentCustomization{}
+		refNamespacedName := types.NamespacedName{Namespace: pool.Namespace, Name: ref.Name}
+		// TODO: context should be TODO or Background?
+		// I'm not sure if this will work
+		err := r.Get(context.TODO(), refNamespacedName, cdc)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				logger.Infof("customization not found: %s", ref.Name)
+			}
+			log.WithError(err).Error("error reading customization")
+		}
+
+		// TODO: kinda fugly
+		currentAvailability := controllerutils.FindClusterDeploymentCustomizationCondition(
+			cdc.Status.Conditions,
+			hivev1.ClusterDeploymentCustomizationAvailableCondition,
+		)
+		// Customization available
+		if currentAvailability.Status == corev1.ConditionFalse {
+			r.setInventoryValidCondition(pool, false, logger)
+			continue
+		}
+		if cdc.Status.ClusterDeploymentRef == nil && currentAvailability.Status == corev1.ConditionTrue {
+			//
+			cdc.Status.Conditions = controllerutils.SetClusterDeploymentCustomizationCondition(
+				cdc.Status.Conditions,
+				hivev1.ClusterDeploymentCustomizationAvailableCondition,
+				corev1.ConditionFalse,
+				"processing",
+				"processing",
+				controllerutils.UpdateConditionIfReasonOrMessageChange, // TODO: maybe always?
+			)
+			r.setInventoryValidCondition(pool, true, logger)
+			return &corev1.LocalObjectReference{Name: cdc.Name}, nil
+		}
+	}
+	return nil, errors.New(fmt.Sprintf("no customization available"))
+}
+
+func (r *ReconcileClusterPool) getCustomizedInstallConfigSecretRef(cd *hivev1.ClusterDeployment, customizationRef *corev1.LocalObjectReference, logger log.FieldLogger) error {
+	// TODO: Am I doing this right?
```

Comment:
Instead of patching yaml secret why don't you patch `installConfigTemplate` before passing it to the builder. This library might help you do it https://github.com/evanphx/json-patch.

Reply:
That was my first implementation, but Builder merges the provided template with its own values or generates a new install config  (https://github.com/openshift/hive/blob/master/pkg/clusterresource/builder.go#L201) and the merge process favorites Builder values( https://github.com/openshift/hive/blob/8dc97bc97ec826dbc84b1ae9c9bc872b323eb0fd/pkg/clusterresource/builder.go#L424) so it will be counter intuative for the user if the patches are overwritten by hive internal logic. That is why I have put the patching process at the very end of ClusterDeployment creation, and it uses `json-patch` (see `applyPatches` function).

Reply:
Only things changed by `mergeInstallConfigTemplate` function are `name` and `basedomain`. Extracting and then patching data from generated installConfigSecret, as currently written, looks like a tedious process to me. But yes theoretically it should work. I would like to get a second opinion here. @2uasimojo @suhanime  any thoughts?